Feat: clang-tidy for linting of C/C++ code

peakschris opened the pull request:
Support clang-tidy for linting of C++ code.

Inspired by and containing some code from bazel_clang_tidy repo. This integration has been discussed by maintainers of both repos in https://github.com/erenon/bazel_clang_tidy/issues/35.

This PR replaces #284, which had become noisy with too many commits. Since then, the code has been rewritten to return a single report for each target.

Fixes #178 

---

### Changes are visible to end-users: yes

- Searched for relevant documentation and updated as needed: yes
- Breaking change (forces users to change their own code or config): no
- Suggested release notes appear below: todo

### Test plan

- New test cases added
- Extra test cases could be added but need discussion; what is possible in CI?

### Code used from bazel_clang_tidy repo

bazel_clang_tidy repo helped:
- two helper functions toolchain_flags and safe_flags
- some lines in get_args
- inspiration & learning

The bulk of the code is implemented by following existing rules_lint code such as ruff.bzl and eslint.bzl.

### Input into release notes (will reformat)
- warning that git bash does not handle arguments correctly and --fix may have issues on Windows. msys64 bash (already recommended by Bazel is ok).

peakschris commented:
I'm fairly happy with this, it's getting closer now...

Supported:
- bash lint.sh //src/...  _(windows)_
- ./lint.sh //src/... _(linux)_
- ./lint.sh --fix //src:hello_cc _(linux)_
- ./lint.sh --fix //src/cpp/... _(linux)_

Needs discussion:
- some todo statements in clang_tidy_aspect_impl, need to decide which targets are ignored
- error code handling in clang_tidy_wrapper.sh, I am distinguishing fatal from user-selected errors
- workaround_windows.bat fixes; can we make example repo support windows by default and remove this?

Issues that I can't fix:
- bash lint.sh --fix //src:hello_cc fails on windows due to missing node.bat: `FATAL: aspect_rules_js[js_binary]: node wrapper '/d/udu/b/l2x2bhay/execroot/_main/./external/aspect_rules_lint~/lint/private/patcher_node_bin/node.bat' not found`
- can't package a clang-tidy.exe on windows as llvm_toolchain doesn't support windows at present

alexeagle commented:
I'm okay landing this without windows support, we don't have testing there currently https://github.com/aspect-build/rules_lint/blob/main/.bcr/presubmit.yml#L5 so there's an existing project to fix it. Not many OSS contributors enthusiastic about both Bazel and Windows because so much is broken, but I hope to help fix that :)

peakschris commented:
@alexeagle thanks, agreed we need separate tracking for windows. Would be good to get this in for linux for now. I've removed the windows_workaround.bat.

Please would you restart CI? I've fixed some of the failures.

peakschris commented on example/lint.sh:
```diff
@@ -59,7 +76,12 @@ fi
 # TODO: Maybe this could be hermetic with bazel run @aspect_bazel_lib//tools:jq or sth
-valid_reports=$(jq --arg ext report --raw-output "$filter" "$buildevents")
+if [ $machine == "Windows" ]; then
+    # jq on windows outputs CRLF which breaks this script. https://github.com/jqlang/jq/issues/92
```
this line fixes #178

alexeagle commented on example/.clang-tidy:
```diff
@@ -0,0 +1,11 @@
+Checks: >
```
nit: can it have a comment pointing to docs?

alexeagle commented on example/lint.sh:
```diff
@@ -22,9 +22,26 @@ fix=""
+
+args=()
+if [ $machine == "Windows" ]; then
+    # avoid missing linters on windows platform
```
nit: you should run `pre-commit install` so this gets indented for you

peakschris replied:
Didn't know about pre-commit hooks, I've learned something new, thank you! The buildifier hooks are failing on windows with missing /bin/bash errors. And all files are getting converted from CRLF to LF, I assume this is not desired?

alexeagle commented on example/src/cpp/lib/hello-time.cc:
```diff
@@ -0,0 +1,10 @@
+#include <iostream>
+#include <get/get-time.h>
+
+// Deliberately bad code
```
in general, comment a bit more about what violation it's expected to produce

alexeagle commented on example/tools/lint/BUILD.bazel:
```diff
@@ -68,3 +70,17 @@ copy_to_directory(
+            "@bazel_tools//src/conditions:linux_aarch64": "@llvm_toolchain_llvm//:bin/clang-tidy",
+            "@bazel_tools//src/conditions:darwin_x86_64": "@llvm_toolchain_llvm//:bin/clang-tidy",
+            "@bazel_tools//src/conditions:darwin_arm64": "@llvm_toolchain_llvm//:bin/clang-tidy",
+            "@bazel_tools//src/conditions:windows_x64": "clang-tidy.exe",
```
comment: link to issue about missing llvm toolchain bits?

alexeagle commented on example/tools/lint/linters.bzl:
```diff
@@ -71,3 +72,12 @@ ktlint = lint_ktlint_aspect(
+
+clang_tidy = lint_clang_tidy_aspect(
+    binary = "@@//tools/lint:clang_tidy",
+    config = "@@//:.clang-tidy",
```
curious, can a subdirectory in a monorepo override some settings for their project? If they have a nested .clang-tidy file, is the tool expected to resolve the nearest-ancestor file, and then do settings from parent directories still apply?

peakschris replied:
A .clang-tidy file in any directory can be specified here, I am specifying this absolute path on the command-line to clang-tidy -- there is no search. It would be interesting to see if I could set it up to allow different configs per-directory.

peakschris replied:
Yes, I've added this support as an option. The setup is rather clumsy, you need to add all the submodule .clang-tidy files to the global aspect. It would be nicer if these could be declared in each nearest BUILD file.

alexeagle commented on lint/clang_tidy_wrapper.sh:
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/bash
+# This is a wrapper for clang-tidy which allows it to be run on a collection of files
```
Could you file and link to an upstream issue about the missing feature upstream? At least that gives us a chance of handing this script to them, and not have to maintain it here.

peakschris replied:
Done, turns out that multiple files are supported, and I missed it. I've now refactored to pass multiple files to clang-tidy. The wrapper is still in the picture at present, distinguishing between errors from checks and fatal errors.

peakschris replied:
I've opened a llvm feature request https://github.com/llvm/llvm-project/issues/95721 to allow distinguishing between fatal and check error through exit code. If that was supported, the wrapper could be removed.

alexeagle commented on lint/clang_tidy_wrapper.sh:
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/bash
+# This is a wrapper for clang-tidy which allows it to be run on a collection of files
+# Usage: clang-tidy-target.bash <clang-tidy-path> <file1> <file2> ... -- <compiler-args>
```
nit: .bash -> .sh? (actually I think `.bash` is the correct extension for this file if it doesn't work under other shells)

peakschris replied:
Fixed

alexeagle commented on lint/clang_tidy.bzl:
```diff
@@ -0,0 +1,337 @@
+    if not CcInfo in target:
+        return []
+
+    # todo: keep this? #Ignore external targets
```
we certainly shouldn't produce lint results for source files that the user doesn't control. Do we need to uncomment this? Or do you need to use a `filter_srcs` helper like the sibling linters?

peakschris replied:
Removed commented blocks, using filter_srcs

alexeagle commented on lint/clang_tidy.bzl:
```diff
@@ -0,0 +1,337 @@
+    #if target.label.workspace_root.startswith("external"):
+    #    return []
+
+    # Targets with specific tags will not be formatted
```
formatted -> linted?

peakschris replied:
Removed this

alexeagle commented on lint/clang_tidy.bzl:
```diff
@@ -0,0 +1,337 @@
+    #    return []
+
+    # Targets with specific tags will not be formatted
+    # todo: does this align with rules_lint framework?
```
No, because the user can choose which targets they'd like to lint. If it's linting something you didn't mean to, in theory you pass a different target pattern as the argument.

That said, `--test_tag_filters` is pretty convenient for testing and we don't have an equivalent for linting.

Anyway I'd just remove this since such a feature belongs in a layer that applies to all languages.

peakschris replied:
Removed

alexeagle commented on lint/clang_tidy.bzl:
```diff
@@ -0,0 +1,337 @@
+            "linter": executable._clang_tidy_wrapper.path,
+            "args": [executable._clang_tidy.path, "--fix"] + _get_args(ctx, compilation_context, srcs),
+            "env": {
+                #"CLANG_TIDY__VERBOSE": "1",
```
comment about why this is commented out?

peakschris replied:
It's for debugging. shall I pass a verbose flag through from aspect factory to control this, to make it easier for integrators to see the clang-tidy command lines being invoked?

peakschris replied:
Added this as user configuration via factory option

alexeagle reviewed (commented):
Awesome, excited to have this. /cc @jsharpe

peakschris commented:
@alexeagle looks like there are 3 ci pipelines failing... are they due to this PR? I couldn't see the relationship between this change and the starlark failures... Can I run the integration tests myself, to see what the issue is there?

peakschris commented:
@alexeagle have pushed a fix for latest ci run

alexeagle commented:
I ran buildifier locally and corrected what it pointed out, see the most recent commit.

peakschris commented:
Awesome, thanks for your help @alexeagle. What are your thoughts? I'm happy to do more work if needed on this, either before or after merge.

alexeagle reviewed (approved):
LGTM let's ship and get some user feedback 👍🏻

peakschris commented:
Amazing! Thank you for your help with this 🥳
